mvrx/src/main/kotlin/com/airbnb/mvrx/MavericksViewModel.kt

```diff
@@ -415,68 +418,72 @@ abstract class MavericksViewModel<S : MavericksState>(
         lifecycleOwner: LifecycleOwner? = null,
         deliveryMode: DeliveryMode,
         action: suspend (T) -> Unit
-    ): Job {
-        val flow = if (lifecycleOwner == null || MavericksTestOverrides.FORCE_DISABLE_LIFECYCLE_AWARE_OBSERVER) {
-            this
-        } else if (deliveryMode is UniqueOnly) {
-            val lastDeliveredValue: T? = lastDeliveredValue(deliveryMode)
-            this
-                .assertOneActiveSubscription(lifecycleOwner, deliveryMode)
-                .dropWhile { it == lastDeliveredValue }
-                .flowWhenStarted(lifecycleOwner)
-                .distinctUntilChanged()
-                .onEach { lastDeliveredStates[deliveryMode.subscriptionId] = it }
-        } else {
-            flowWhenStarted(lifecycleOwner)
-        }
+    ): Job = collectLatest(this@MavericksViewModel, lifecycleOwner, deliveryMode, action)
 
-        val scope = (lifecycleOwner?.lifecycleScope ?: viewModelScope) + configFactory.subscriptionCoroutineContextOverride
-        return scope.launch(start = CoroutineStart.UNDISPATCHED) {
-            // Use yield to ensure flow collect coroutine is dispatched rather than invoked immediately.
-            // This is necessary when Dispatchers.Main.immediate is used in scope.
-            // Coroutine is launched with start = CoroutineStart.UNDISPATCHED to perform dispatch only once.
-            yield()
-            flow.collectLatest(action)
+    fun <T> lastDeliveredValue(subscriptionId: String?): T? {
+        @Suppress("UNCHECKED_CAST")
+        return subscriptionId?.let { lastDeliveredStates[it] } as T?
+    }
+
+    private fun <S : MavericksState> assertSubscribeToDifferentViewModel(viewModel: MavericksViewModel<S>) {
+        require(this != viewModel) {
+            "This method is for subscribing to other view models. Please pass a different instance as the argument."
         }
     }
 
-    @Suppress("EXPERIMENTAL_API_USAGE")
-    private fun <T> Flow<T>.assertOneActiveSubscription(owner: LifecycleOwner, deliveryMode: UniqueOnly): Flow<T> {
-        val observer = object : DefaultLifecycleObserver {
-            override fun onCreate(owner: LifecycleOwner) {
-                if (activeSubscriptions.contains(deliveryMode.subscriptionId)) error(duplicateSubscriptionMessage(deliveryMode))
-                activeSubscriptions += deliveryMode.subscriptionId
-            }
+    override fun toString(): String = "${this::class.java.simpleName} $state"
+}
 
-            override fun onDestroy(owner: LifecycleOwner) {
-                activeSubscriptions.remove(deliveryMode.subscriptionId)
-            }
+fun <T : Any, S: MavericksState> Flow<T>.collectLatest(
+    viewModel: MavericksViewModel<S>,
```

Comment:
I  think we can actually remove the viewmodel parameter from this once the function is moved to the MavericksView. The big benefit to this is it removes the weirdness around the expectation that the viewmodel you use is the same one that the state flow comes from, and creates a nicer API.

The viewmodel right now is used to store the active subscriptions and last value - that can instead live in the mavericks view via the same pattern we use for the mvrx view id
```
val mvrxViewId: String
        get() = when (this) {
            is ViewModelStoreOwner -> ViewModelProvider(this).get(MavericksViewIdViewModel::class.java).mavericksViewId
```

The viewmodel is also used for the config factory for the coroutine context override, but that can be accessed statically via `Mavericks.viewModelConfigFactory`?

We would then also have to refactor this function to be more generic to work for both the mavericksview case and the old subscription case inside the viewmodel, but we can help do that by creating a new class to hold the activeSubscriptions and lastDeliveredStates values and manage them.

What do you think? I know this is even more work, but should make the API quite a bit nicer and would be a hard thing to change later